aten/src/ATen/cpp_custom_type_hack.h

```diff
@@ -24,23 +26,45 @@ T& cast(const Tensor& packed) {
   return *reinterpret_cast<T*>(packed.storage().data_ptr().get());
 }
 
-template<typename T>
+template <typename T>
 Tensor create(std::unique_ptr<T> ptr, TensorOptions options) {
   // We store this instance away in a Tensor and register a deleter function
   // so that we do not leak memory. On the other side, we pull out the storage's
   // data_ptr and get the right typed pointer.
   void* raw_ptr = ptr.release();
   at::DataPtr at_ptr(
-      raw_ptr,
-      raw_ptr,
-      caffe2::TypeMeta::Make<T>().deleteFn(),
-      at::kCPU);
+      raw_ptr, raw_ptr, caffe2::TypeMeta::Make<T>().deleteFn(), at::kCPU);
 
   // size doesn't really matter, but we can align it to the actual size
   // returning variables because one likely want to use this hack from python
   auto retval = at::empty({sizeof(T)}, options.device(kCPU).dtype(at::kByte));
   retval.storage().set_data_ptr(std::move(at_ptr));
   return retval;
 }
+
+template <typename T>
+QTensor create(
```

Comment:
why do you need it? - you can just return normal CPUTensor as before (as it's a hack). If you're worried about dispatch - you'd be fine, the first argument is X which is a proper QTensor (and in the worst case we can just remove dispatch all together as the op is called explicitly quantized::fc)

Reply:
@dzhulgakov: Sorry just saw this comment. We want the ```create``` function to return an at::QTensor (now it is renamed as at::Tensor)  with scale and zero_point information for the prepacked weight. If we use the original ```create``` function, it returns a normal at:Tensor, which is not consistent with the type for input activation and output tensor (at:QTensor).

Reply:
I can add ```scale``` and ```zero_point``` inside the special struct of ```PackedFCWeight``` in ```fbgemm_utils.h```, then the prepacked weight will be a normal ByteTensor. In this way we don't need this function. Should we do this?

Reply:
I don't think it's a good idea to use QTensor like this - please make it regular ByteTensor. The fact that it's a tensor at all is temporary - we will replace it with opaque custom type soon. Also, bundling scale and bias in slots in the QTensor doesn't make much sense as you won't be able to perform any other ops on it.

Also, do you even need to pack scale and zero_point? If I read you code correctly, you don't ever use them

Reply:
@dzhulgakov: Got it. I actually need to use ```weight_scale``` and ```weight_zero_point```. ```weight_scale``` is used to compute ```output_multiplier_float```.  ```weight_zero_point``` and ```output_multiplier_float``` are needed passed into ```fbgemm::ReQuantizeOutput``` requantization function. Of course, weight_scale can be folded into output_scale in advance, but ```weight_zero_point``` is required for the requantization. We need to multiply ```weight_zero_point``` with the ```row_offset``` (the sum of the rows of the input activation A matrix) and add that result to the quantized input/weight product during the requantization step.